### PR TITLE
First pass at MonoClass reorg

### PR DESCRIPTION
**Pull request by kumpera**
Still lots more to come. Early feedback appreciated.

### CODE REVIEW

File: mono/metadata/class-inlines.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
+/* 
+ * Copyright 2014 Xamarin Inc
+ */
+#ifndef __MONO_METADATA_CLASS_INLINES_H__
+#define __MONO_METADATA_CLASS_INLINES_H__
+
+#include <mono/metadata/class-internals.h>
+
+static inline gboolean
+mono_class_is_def (MonoClass *class)
```
Comment (evincarofautumn) [inline mono/metadata/class-inlines.h]: We should probably use `klass` for C++ compat.

**Comment by kumpera** [comment]
From the above feedback:
1. Add flag test functions/macros
2. Add missing license headers
3. Try to remove the has_generic_params call from mono_create_from_typedef

**Comment by kumpera** [comment]
build

**Comment by kumpera** [comment]
This is now ready for merging

File: mono/metadata/class.c [hunk 2]
```diff
@@ -6719,11 +6716,10 @@ mono_bounded_array_class_get (MonoClass *eclass, guint32 rank, gboolean bounded)
 
 	mono_profiler_class_event (klass, MONO_PROFILE_START_LOAD);
 
-	classes_size += sizeof (MonoClass);
+	classes_size += sizeof (MonoClassArray);
+	++class_array_count;\
```
Comment (Therzok) [inline mono/metadata/class.c]: is this backslash intended?

**Comment by henricm** [comment]
This PR breaks the windows build. Created https://github.com/mono/mono/pull/3893 to fix that.

**Pull request merged**